src/OrleansRuntime/GrainDirectory/LocalGrainDirectory.cs

```diff
@@ -755,6 +794,33 @@ public List<ActivationAddress> GetLocalCacheData(GrainId grain)
                 null;
         }
 
+        public Task<AddressesAndTag> LookupInCluster(GrainId grainId, string clusterId)
+        {
+            if (clusterId == null)
+                throw new ArgumentNullException("clusterId");
+
+            if (clusterId == ClusterId)
+            {
+                return LookupAsync(grainId);
+            }
+            else
+            {
+                // find gateway
+                var gossipOracle = Silo.CurrentSilo.LocalMultiClusterOracle;
+                var clusterGatewayAddress = gossipOracle.GetRandomClusterGateway(clusterId);
```

Comment:
I have updated `GetRandomClusterGateway`  to be more efficient as described above.